File: javascript/clock/clock.js
var minutesPerHour = 60;
var hoursPerDay = 24;

var Clock = function(hour, minute) {
  this.hour = hour || 0;
  this.minute = minute || 0;
};

Clock.prototype.toString = function() {
  this.rollOverMinutesToHours();
  var hour = this.adjustTime(this.hour, hoursPerDay);
  var minute = this.adjustTime(this.minute, minutesPerHour);
  return this.formatTime(hour, minute);
};

Clock.prototype.plus = function(minutes) {
  this.minute += minutes;
  return this;
};

Clock.prototype.minus = function(minutes) {
  this.minute -= minutes;
  return this;
};

Clock.prototype.equals = function(clock) {
  return this.toString() === clock.toString();
};

Clock.prototype.rollOverMinutesToHours = function() {
  this.hour += Math.floor(this.minute / minutesPerHour);
};

Clock.prototype.adjustTime = function(time, adjustor) {
  var adjustedTime = time % adjustor;
  if (adjustedTime < 0) { adjustedTime += adjustor; }
  return adjustedTime;
};

Clock.prototype.zeroPad = function(time) {
  if (time < 10) { time = "0" + time; }
  return time;
};

Clock.prototype.formatTime = function(hour, minute) {
  return this.zeroPad(hour) + ":" + this.zeroPad(minute);
};

function at(hour, minute) {
  return new Clock(hour, minute);
};

module.exports = {at: at};

File: ruby/clock/clock.rb
class Clock
  HOURS = 24
  MINUTES = 60

  def self.at(hour, minute)
    new(hour, minute)
  end

  def to_s
    adjust_time
    "#{zero_pad(hour)}:#{zero_pad(minute)}"
  end

  def +(other)
    @minute += other
    self
  end

  def -(other)
    @minute -= other
    self
  end

  def ==(other)
    to_s == other.to_s
  end

  private

  attr_reader :hour, :minute

  def initialize(hour, minute)
    @hour = hour
    @minute = minute
  end

  def zero_pad(num)
    num.to_s.rjust(2, '0')
  end

  def adjust_hour
    @hour = hour % HOURS
  end

  def adjust_minute
    @hour += (minute / MINUTES)
    @minute = minute % MINUTES
  end

  def adjust_time
    adjust_minute
    adjust_hour
  end
end

module BookKeeping
  VERSION = 2
end
